File: test/native/flimsgfrom.cpp
#include <dragon/messages.hpp>
#include <dragon/fli.h>
#include <dragon/utils.h>
#include <iostream>

using namespace std;

int main(int argc, char* argv[]) {
    try {
        dragonError_t err;
        dragonFLIDescr_t fli;
        dragonFLISendHandleDescr_t sendh;
        dragonFLISerial_t ser_fli;

        if (argc != 2) {
            cout << "FAILED: This program expects a serialized fli as its command-line argument." << endl;
            return -1;
        }

        const char* serFLIb64 = argv[1];
        ser_fli.data = dragon_base64_decode(serFLIb64, &ser_fli.len);

        err = dragon_fli_attach(&ser_fli, NULL, &fli);
        if (err != DRAGON_SUCCESS) {
            cout << "FAILED to attach to FLI" << endl;
            cerr << "FAILED to attach to FLI" << endl;
            return -1;
        }

        err = dragon_fli_open_send_handle(&fli, &sendh, NULL, NULL, false, true, NULL);
        if (err != DRAGON_SUCCESS) {
            cout << "Failed to open send handle" << endl;
            cerr << "Failed to open send handle" << endl;
            return -1;
        }

        DDRegisterClientMsg msg(42, "Hello World", "Dragon is the best");

        msg.send(&sendh, NULL);

        cout << "OK" << endl;
        return 0;
    } catch (...) {
        cerr << "Exception in code." << endl << flush;
        cout << "Exception in code." << endl << flush;
    }

    return -1;
}

File: test/native/flimsgfrom.py
#!/usr/bin/env python3

import unittest
import os
import dragon
import sys
import multiprocessing as mp
from dragon.fli import FLInterface, DragonFLIError, DragonFLIEOT
from dragon.managed_memory import MemoryPool, MemoryAlloc
from dragon.channels import Channel
from dragon.globalservices import channel
from dragon.localservices.options import ChannelOptions
from dragon.native.process import Popen
import dragon.infrastructure.messages as dmsg
import dragon.infrastructure.facts as facts
import dragon.infrastructure.parameters as parameters
from dragon.utils import b64decode, b64encode


def main():
    try:
        ser_fli = sys.argv[1]
        decoded_ser_fli = b64decode(ser_fli)
        fli = FLInterface.attach(decoded_ser_fli)
        msg = dmsg.DDRegisterClient(42, "Hello World", "Dragon is the best")
        sendh = fli.sendh()
        sendh.send_bytes(msg.serialize())
        sendh.close()
        print("OK")
    except Exception as ex:
        print(f"Got Exception in flimsgfrom.py {ex}", flush=True, file=sys.stderr)


if __name__ == "__main__":
    main()
